modal/client.py

```diff
@@ -226,44 +225,36 @@ async def from_env(cls, _override_config=None) -> "_Client":
         else:
             c = config
 
-        server_url = c["server_url"]
-
-        token_id = c["token_id"]
-        token_secret = c["token_secret"]
-        task_id = c["task_id"]
-        credentials = None
-
-        if task_id:
-            client_type = api_pb2.CLIENT_TYPE_CONTAINER
-        else:
-            client_type = api_pb2.CLIENT_TYPE_CLIENT
-            if token_id and token_secret:
-                credentials = (token_id, token_secret)
-
         if cls._client_from_env_lock is None:
             cls._client_from_env_lock = asyncio.Lock()
 
         async with cls._client_from_env_lock:
             if cls._client_from_env:
                 return cls._client_from_env
+
+            server_url = c["server_url"]
+
+            if _is_remote():
+                credentials = None
+                client_type = api_pb2.CLIENT_TYPE_CONTAINER
```

Comment:
I wonder if we're setting ourselves up for (more) confusion about the two distinct meanings of the local/remote dichotomy in Modal:

- practical sense:
  - local: not on modal infrastructure
  - remote: on modal infrastructure
- functional sense:
  - local: the context that you run or deploy your applications _from_
  - remote: the context that you run or deploy your applications _in_

(This is maybe not a perfect set of definitions but hopefully it makes sense)

The `_is_remote` helper indicates which side you're on in the practical sense (based on the environment variables that we define in containers).

But what about users who want to run or deploy applications from Modal infrastructure? I thought we had people doing that; including people doing it on behalf of their users (such that the orchestrator app and worker app are using different tokens). Are we breaking that behavior? Maybe we're not — but it feels very hard to reason about.

Reply:
Let me add a test for deploying Modal apps from inside containers. It's possible this PR broke it because of the `authenticated` check. That being said, the authentication system (client vs container vs web) is basically orthogonal to what we're doing (creating an app vs running code inside an app).